src/settings.ts

```diff
@@ -417,7 +417,7 @@ export function makeRootDatabaseConnectionString(
       return `socket:${parsed.pathname}?${query}`;
     }
   } else {
-    parsed.pathname = `/${databaseName}`;
+    parsed.pathname = databaseName;
```

Comment:
My first gut reaction was that I'd rather test it thoroughly at dev time than introduce a branch at runtime, but the docs on url show the pathname going in and coming out always with a leading slash. The docs make no reference to the leading slash's behavior as a part of the API, so, I think you're right here, that it would be risky.

https://nodejs.org/api/url.html#urlpathname

+ I also realized that since url is a core library, the behavior of url is not encoded in package.json / the lock file, so the dependency could change out from under us without a CI run if it's broken in a later node version.

Interestingly, url's `parse()` treats the string "somedb" as, in fact, a path. I'm not familiar enough with the formal URL spec here to know if the library is behaving according to spec, but I'm assuming it is. It sounds like we can't rely on the case where the `url` lib thinks of the string as a non-url.

```
Welcome to Node.js v16.13.2.
Type ".help" for more information.
> require('url').parse('somedb');
Url {
  protocol: null,
  slashes: null,
  auth: null,
  host: null,
  port: null,
  hostname: null,
  hash: null,
  search: null,
  query: null,
  pathname: 'somedb',
  path: 'somedb',
  href: 'somedb'
}
> require('url').format(require('url').parse('somedb'));
'somedb'
> require('url').format(require('url').parse('/somedb'));
'/somedb'
```

Reply:
Apologies for the word vomit.

I read through the URI RFC, and I thought it made sense to trust that the url library will reliably use the path as a delimiter between the host component and path component of the URI. If it didn't, it would be out-of-spec-ish with the RFC.

[Section 1.2.3](https://www.rfc-editor.org/rfc/rfc3986#section-1.2.3) says that `For some URI schemes, the visible hierarchy is limited to the scheme itself: everything after the scheme component delimiter (":") is considered opaque to URI processing.  Other URI schemes make the hierarchy explicit and visible to generic parsing algorithms. The generic syntax uses the slash ("/"), question mark ("?"), and number sign ("#") characters to delimit components that are significant to the generic parser's hierarchical interpretation of an identifier`

This certainly implies that the slash aught to be used as a delimiter, but it doesn't explicitly say it must.

[Section 3.2](https://www.rfc-editor.org/rfc/rfc3986#section-3.2) says `If a URI contains an authority component, then the path component must either be empty or begin with a slash ("/") character` this is the important line. This suggests to me that if the format function did not prepend a slash on a path when there is an authority of any kind, it would be violating the spec.

So this suggests we're good to rely on the slash, and yet...

```
> require('url').format({ pathname: 'foobar', hostname: 'abc.com', protocol: 'postgres' });
'postgres:abc.comfoobar'
> require('url').format({ pathname: 'foobar', hostname: 'abc.com', protocol: null, auth: 'user:pass' });
'user:pass@abc.comfoobar'
> require('url').format({ pathname: 'foobar', hostname: 'abc.com', protocol: 'postgres', slashes: true });
'postgres://abc.com/foobar'
> require('url').format({ pathname: 'foobar', hostname: 'abc.com', protocol: null });
'abc.comfoobar'
> require('url').format({ pathname: 'foobar', hostname: 'abc.com', protocol: null, slashes: true });
'//abc.com/foobar'
```

I did test that when `slashes: true`, it will always use a slash as delimiter between hostname and path, and that if the url contains `//` between the scheme and the hostname, then calling parse() will always return a url object with `slashes: true`.  This specific behavior of parse is unfortunately undocumented, and thus we fall into the same risk of it being susceptible to unexpected breaking changes.

The behavior above is only out-of-spec-ish because the URI RFC basically says "meh, schemes can do kinda whatever they want. They don't necessarily have to respect the generic parsing algorithms"

So... let's go see what postgres has to say about this. See `33.1.1.2. Connection URIs` in [docs](https://www.postgresql.org/docs/12/libpq-connect.html). It specifically says that connection URIs must start with `postgres` or `postgresql`, and it contains no documentation suggesting that "connection strings" can come in the form of a standalone db name.

[Node-postgres](https://node-postgres.com/features/connecting) specifically references Connection URIs. The docs do not suggest that `somedbname` is a valid connection uri. But if I try it... it works

```
> const { Pool } = require('pg')
undefined
> new Pool({ connectionString: 'template1' }).query('select now()', console.log);
undefined
> undefined Result {
  command: 'SELECT',
  rowCount: 1,
  oid: null,
  rows: [ { now: 2022-03-01T07:11:29.109Z } ],
  ...
}
```

I'm guessing this is just a psql convention that node-postgres adopted. [Psql docs](https://www.postgresql.org/docs/13/app-psql.html). `If an argument is found that does not belong to any option it will be interpreted as the database name`

I'm assuming graphile migrate would want to follow node-postgres's lead, even if it's not well documented. If so, I'd recommend adding a regex that checks if it's a valid db name, and if so, treat it differently.

Unfortunately... there are no restrictions for postgres database names beyond those of identifiers, which means you can do a lot including unicode characters 😳 

```
create database U&"\0441\043B\043E\043D";
CREATE DATABASE
Time: 0.177s
```

This works!

```
> new Pool({ connectionString: '\u0441\u043B\u043E\u043D' }).query('select now()', console.log);
undefined
> undefined Result {
  command: 'SELECT',
  rowCount: 1,
  oid: null,
  rows: [ { now: 2022-03-01T07:23:42.934Z } ],
  ...
}
```

Do we really want to go down this path?

(Wow the 🐰 🕳️ is real when trying to protect against people using a library in unexpected ways)

Reply:
Can we just gate it on `parsed.protocol`?

Reply:
... yep. That seems like that will work 😅 

Updated function